File: examples/go/agg-table/handlers/end_hand.go
package handlers

import (
	"encoding/hex"
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// HandleEndHand handles the EndHand command.
func HandleEndHand(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state TableState,
	seq uint32,
) (*pb.EventBook, error) {
	if !state.Exists() {
		return nil, angzarr.NewCommandRejectedError("Table does not exist")
	}

	var cmd examples.EndHand
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	// Verify we're in a hand
	if state.Status != "in_hand" {
		return nil, angzarr.NewCommandRejectedError("No hand in progress")
	}

	// Verify hand root matches
	if hex.EncodeToString(cmd.HandRoot) != hex.EncodeToString(state.CurrentHandRoot) {
		return nil, angzarr.NewCommandRejectedError("Hand root mismatch")
	}

	// Calculate stack changes from pot results
	stackChanges := make(map[string]int64)
	for _, result := range cmd.Results {
		winnerHex := hex.EncodeToString(result.WinnerRoot)
		stackChanges[winnerHex] += result.Amount
	}

	event := &examples.HandEnded{
		HandRoot:     cmd.HandRoot,
		Results:      cmd.Results,
		StackChanges: stackChanges,
		EndedAt:      timestamppb.New(time.Now()),
	}

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}
	eventAny.TypeUrl = "type.poker/examples.HandEnded"

	return &pb.EventBook{
		Cover: commandBook.Cover,
		Pages: []*pb.EventPage{
			{
				Sequence:  &pb.EventPage_Num{Num: seq},
				Event:     eventAny,
				CreatedAt: timestamppb.New(time.Now()),
			},
		},
	}, nil
}

File: examples/rust/agg-table/src/handlers/end_hand.rs
//! EndHand command handler.

use std::collections::HashMap;

use angzarr_client::proto::examples::{EndHand, HandEnded};
use angzarr_client::proto::{event_page, CommandBook, EventBook, EventPage};
use angzarr_client::{pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::TableState;

pub fn handle_end_hand(
    command_book: &CommandBook,
    command_any: &Any,
    state: &TableState,
    seq: u32,
) -> CommandResult<EventBook> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Table does not exist"));
    }

    let cmd: EndHand = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    // Verify we're in a hand
    if state.status != "in_hand" {
        return Err(CommandRejectedError::new("No hand in progress"));
    }

    // Verify hand root matches
    if hex::encode(&cmd.hand_root) != hex::encode(&state.current_hand_root) {
        return Err(CommandRejectedError::new("Hand root mismatch"));
    }

    // Calculate stack changes from pot results
    let mut stack_changes: HashMap<String, i64> = HashMap::new();
    for result in &cmd.results {
        let winner_hex = hex::encode(&result.winner_root);
        *stack_changes.entry(winner_hex).or_insert(0) += result.amount;
    }

    let event = HandEnded {
        hand_root: cmd.hand_root,
        results: cmd.results,
        stack_changes,
        ended_at: Some(angzarr_client::now()),
    };

    let event_any = pack_event(&event, "examples.HandEnded");

    Ok(EventBook {
        cover: command_book.cover.clone(),
        pages: vec![EventPage {
            sequence: Some(event_page::Sequence::Num(seq)),
            event: Some(event_any),
            created_at: Some(angzarr_client::now()),
        }],
        snapshot: None,
        next_sequence: 0,
    })
}
